veps/sig-storage/vm-object-graph.md

```diff
@@ -0,0 +1,279 @@
+# VEP #41: Object Graph API for VM Dependencies
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] : https://github.com/kubevirt/enhancements/issues/41
+
+## **Overview**
+
+This is a proposal to include an `Object Graph` API in KubeVirt to represent VM and VMI dependencies and their relationships.
+
+## **Motivation**
+
+As new features continue to be added to KubeVirt, the graph of objects related to VMs (DataVolumes, PersistentVolumeClaims, InstanceTypes, Preferences, Secrets, ConfigMaps, etc.) continues to expand. Identifying all the objects that a VM depends on for tasks like backup, disaster recovery, or migration can be error-prone. We should simplify this process for users and partners by creating an authoritative way to retrieve a structured object graph.
+
+## **Goals**
+
+- Introduce a new Object Graph API to represent the list of resources a VM or VMI depends on.
+- Expose this API as a subresource of VirtualMachines and VirtualMachineInstances.
+- Provide a flexible and extensible data structure.
+- Allow for basic filtering (present or future).
+
+## **Non-Goals**
+
+- Reimplementation of existing VM/VMI specs.
+- Building a generic Kubernetes-wide graph system.
+
+## **User Stories**
+
+1. As a KubeVirt user, I want a clear way to retrieve all VM and VMI-related dependencies.
+2. As a backup partner, I want a way to identify a list of a VM's related objects so I can comprehensively backup and restore everything a VM needs.
+3. As a VM owner, I want to easily define an ACM-discovered application and protect my VM with disaster recovery software.
+4. As a VM owner, I want to migrate my VM from one cluster to another and identify all necessary dependencies for replication.
+5. As a KubeVirt developer, I want a specific place to keep the object graph code updated when I introduce code that changes the relationship of a VM to its dependent objects.
+
+## **Repos**
+
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+- [KubeVirt Velero Plugin](https://github.com/kubevirt/kubevirt-velero-plugin)
+
+## **Design**
+
+### API Endpoints
+
+```
+/apis/subresources.kubevirt.io/v1/namespaces/{namespace}/virtualmachines/{name}/objectgraph
+/apis/subresources.kubevirt.io/v1/namespaces/{namespace}/virtualmachineinstances/{name}/objectgraph
```

Comment:
Do we need to define this also for VMIs? Can we just focus on VMs?

Reply:
Are there any downsides to enabling the subresource for VMIs? I’d argue it’s equally important to have it enabled for both VMs and VMIs... for example, when backing up a VMI directly using Velero, we still need to get its object graph.

Reply:
Why are they backing up VMIs? VMIs are ephemeral objects.

Reply:
I think it makes sense to support standalone VMIs. They are backed up with Velero for example

Reply:
Still, it isn't a reason. Why Velero is doing this? I mean if backup VMs and VMIs, then is a VM backed up twice?

Reply:
>  I mean if backup VMs and VMIs, then is a VM backed up twice?

The velero plugin for kubevirt already handles this scenario and ensures only the required resources are backed up for a VM/VMI to run properly, but even if a VM were backed up twice, velero discards duplicates automatically.

> Why Velero is doing this? 

For example, if the backup is being used purely for metadata or debugging purposes (which I believe some customers are doing), it still makes sense to support backing up a standalone VMI imo.